File: android/src/main/kotlin/com/oneviw/oneviw_flutter_sdk/OneviwFlutterSdkPlugin.kt
package com.oneviw.oneviw_flutter_sdk

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Bundle
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Reads OneViw configuration from `<meta-data>` entries declared in the host
 * app's AndroidManifest.xml:
 *
 *   <meta-data android:name="oneviw.PROJECT_TOKEN" android:value="..." />
 *   <meta-data android:name="oneviw.HOST"           android:value="https://your-oneviw-host" />
 *   <meta-data android:name="oneviw.DEBUG"         android:value="true" />   (optional)
 *   <meta-data android:name="oneviw.DISABLE_ATTRIBUTION" android:value="false" />   (optional)
 *   <meta-data android:name="oneviw.REGISTER_CAMPAIGN_SUPER_PROPERTIES" android:value="true" />   (optional)
 */
class OneviwFlutterSdkPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    private companion object {
        const val CHANNEL_NAME = "oneviw_flutter_sdk"
        const val KEY_PROJECT_TOKEN = "oneviw.PROJECT_TOKEN"
        const val KEY_HOST = "oneviw.HOST"
        const val KEY_DEBUG = "oneviw.DEBUG"
        const val KEY_DISABLE_ATTRIBUTION = "oneviw.DISABLE_ATTRIBUTION"
        const val KEY_REGISTER_CAMPAIGN_SUPER_PROPERTIES =
            "oneviw.REGISTER_CAMPAIGN_SUPER_PROPERTIES"
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getNativeConfig" -> result.success(readNativeConfig())
            else -> result.notImplemented()
        }
    }

    private fun readNativeConfig(): Map<String, Any?> {
        val config = HashMap<String, Any?>()
        val metaData = applicationMetaData() ?: return config

        metaData.getString(KEY_PROJECT_TOKEN)?.let { config["projectToken"] = it }
        metaData.getString(KEY_HOST)?.let { config["host"] = it }
        if (metaData.containsKey(KEY_DEBUG)) {
            config["debug"] = metaData.getBoolean(KEY_DEBUG, false)
        }
        if (metaData.containsKey(KEY_DISABLE_ATTRIBUTION)) {
            config["disableAttribution"] =
                metaData.getBoolean(KEY_DISABLE_ATTRIBUTION, false)
        }
        if (metaData.containsKey(KEY_REGISTER_CAMPAIGN_SUPER_PROPERTIES)) {
            config["registerCampaignSuperProperties"] =
                metaData.getBoolean(KEY_REGISTER_CAMPAIGN_SUPER_PROPERTIES, true)
        }
        return config
    }

    private fun applicationMetaData(): Bundle? {
        return try {
            val appInfo: ApplicationInfo = context.packageManager.getApplicationInfo(
                context.packageName,
                PackageManager.GET_META_DATA,
            )
            appInfo.metaData
        } catch (e: PackageManager.NameNotFoundException) {
            null
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: ios/Classes/OneviwFlutterSdkPlugin.swift
import Flutter
import Foundation
import WebKit

/// Reads OneViw configuration from the host app's Info.plist:
///
///   oneviw.PROJECT_TOKEN : String
///   oneviw.HOST           : String
///   oneviw.DEBUG         : Bool   (optional)
///   oneviw.DISABLE_ATTRIBUTION : Bool   (optional)
///   oneviw.REGISTER_CAMPAIGN_SUPER_PROPERTIES : Bool   (optional)
///
/// Also performs the iOS vlink attribution request over an off-screen
/// `WKWebView` (the `vlinkAttribute` method call) so it carries WebKit's
/// User-Agent, cookies and TLS for accurate attribution correlation.
public class OneviwFlutterSdkPlugin: NSObject, FlutterPlugin, WKScriptMessageHandler,
    WKNavigationDelegate
{
    private static let channelName = "oneviw_flutter_sdk"
    private static let keyProjectToken = "oneviw.PROJECT_TOKEN"
    private static let keyHost = "oneviw.HOST"
    private static let keyDebug = "oneviw.DEBUG"
    private static let keyDisableAttribution = "oneviw.DISABLE_ATTRIBUTION"
    private static let keyRegisterCampaignSuperProperties =
        "oneviw.REGISTER_CAMPAIGN_SUPER_PROPERTIES"

    private static let messageHandlerName = "oneviw"
    private static let vlinkTimeout: TimeInterval = 60
    // Error code returned when the WKWebView attempted the request but failed
    // (timeout, navigation error, or in-page fetch error). Dart distinguishes
    // this from a missing native handler and treats it as transient.
    private static let vlinkErrorCode = "vlink_failed"

    // Single-slot state for the in-flight vlink request. Install attribution
    // fires once per cold start (deduped in Dart), so concurrency isn't a
    // real concern; a second concurrent call falls back to plain HTTP (nil).
    private var webView: WKWebView?
    private var pendingResult: FlutterResult?
    private var timeoutTimer: Timer?
    private var fetchUrl = ""
    private var authorization = ""

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        let instance = OneviwFlutterSdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getNativeConfig":
            result(readNativeConfig())
        case "vlinkAttribute":
            guard let args = call.arguments as? [String: Any],
                let url = args["url"] as? String,
                let auth = args["authorization"] as? String
            else {
                result(nil)
                return
            }
            startVlink(url: url, authorization: auth, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Native config

    private func readNativeConfig() -> [String: Any] {
        var config: [String: Any] = [:]
        let info = Bundle.main.infoDictionary ?? [:]

        if let token = info[OneviwFlutterSdkPlugin.keyProjectToken] as? String {
            config["projectToken"] = token
        }
        if let host = info[OneviwFlutterSdkPlugin.keyHost] as? String {
            config["host"] = host
        }
        if let debug = info[OneviwFlutterSdkPlugin.keyDebug] as? Bool {
            config["debug"] = debug
        }
        if let disableAttribution =
            info[OneviwFlutterSdkPlugin.keyDisableAttribution] as? Bool
        {
            config["disableAttribution"] = disableAttribution
        }
        if let registerCampaignSuperProperties =
            info[OneviwFlutterSdkPlugin.keyRegisterCampaignSuperProperties] as? Bool
        {
            config["registerCampaignSuperProperties"] = registerCampaignSuperProperties
        }
        return config
    }

    // MARK: - vlink WKWebView request

    private func startVlink(
        url: String, authorization: String, result: @escaping FlutterResult
    ) {
        // Already busy — let Dart fall back to plain HTTP.
        if pendingResult != nil {
            result(nil)
            return
        }
        guard let endpoint = URL(string: url),
            let scheme = endpoint.scheme,
            let host = endpoint.host
        else {
            result(nil)
            return
        }

        pendingResult = result
        fetchUrl = url
        self.authorization = authorization

        let contentController = WKUserContentController()
        contentController.add(self, name: OneviwFlutterSdkPlugin.messageHandlerName)
        let configuration = WKWebViewConfiguration()
        configuration.userContentController = contentController

        let webView = WKWebView(frame: .zero, configuration: configuration)
        webView.navigationDelegate = self
        self.webView = webView

        // Load minimal HTML with baseURL = endpoint origin so the injected
        // fetch is same-origin and bypasses CORS.
        var origin = "\(scheme)://\(host)"
        if let port = endpoint.port {
            origin += ":\(port)"
        }
        let baseURL = URL(string: origin)
        webView.loadHTMLString(
            "<!doctype html><html><head></head><body></body></html>",
            baseURL: baseURL
        )

        timeoutTimer = Timer.scheduledTimer(
            withTimeInterval: OneviwFlutterSdkPlugin.vlinkTimeout,
            repeats: false
        ) { [weak self] _ in
            self?.finish(with: nil)
        }
    }

    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        webView.evaluateJavaScript(
            buildInjectedFetch(url: fetchUrl, authorization: authorization),
            completionHandler: nil
        )
    }

    public func webView(
        _ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error
    ) {
        finish(with: nil)
    }

    public func webView(
        _ webView: WKWebView,
        didFailProvisionalNavigation navigation: WKNavigation!,
        withError error: Error
    ) {
        finish(with: nil)
    }

    public func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard message.name == OneviwFlutterSdkPlugin.messageHandlerName else { return }
        finish(with: message.body as? String)
    }

    /// Resolve the pending Flutter result once and tear down the WebView.
    ///
    /// A successful body is returned as-is. A `nil` body (timeout / navigation
    /// failure) or the `__oneviw_error` sentinel posted by the injected script
    /// means the WebView ran but failed — that is surfaced as a `FlutterError`
    /// so Dart treats it as a transient failure (no HTTP fallback, no
    /// "attributed" mark, retry next launch) rather than "WebView unavailable".
    private func finish(with body: String?) {
        timeoutTimer?.invalidate()
        timeoutTimer = nil

        let result = pendingResult
        pendingResult = nil

        if let webView = webView {
            webView.configuration.userContentController.removeScriptMessageHandler(
                forName: OneviwFlutterSdkPlugin.messageHandlerName
            )
            webView.navigationDelegate = nil
        }
        webView = nil

        if let body = body, !body.contains("__oneviw_error") {
            result?(body)
        } else {
            result?(
                FlutterError(
                    code: OneviwFlutterSdkPlugin.vlinkErrorCode,
                    message: "vlink WKWebView request failed",
                    details: nil
                )
            )
        }
    }

    private func buildInjectedFetch(url: String, authorization: String) -> String {
        let safeUrl = jsEscape(url)
        let safeAuth = jsEscape(authorization)
        return """
            (function() {
              try {
                fetch("\(safeUrl)", { method: 'GET', headers: { "Authorization": "\(safeAuth)" } })
                  .then(function(r) { return r.text(); })
                  .then(function(t) { window.webkit.messageHandlers.oneviw.postMessage(t); })
                  .catch(function(e) {
                    window.webkit.messageHandlers.oneviw.postMessage(
                      JSON.stringify({ __oneviw_error: String(e && e.message || e) })
                    );
                  });
              } catch (e) {
                window.webkit.messageHandlers.oneviw.postMessage(
                  JSON.stringify({ __oneviw_error: String(e && e.message || e) })
                );
              }
              true;
            })();
            """
    }

    private func jsEscape(_ s: String) -> String {
        return s
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\r")
    }
}
